File: FunnyAlgorithms-master/Bankers_Algorithm_Operating_System/Bankers.java

public class Bankers 
{ 
	int n = 5; // Number of processes 
	int m = 3; // Number of resources 
	int need[][] = new int[n][m]; 
	int [][]max; 
	int [][]alloc; 
	int []avail; 
	int safeSequence[] = new int[n]; 

	void initializeValues() 
	{ 
	// P0, P1, P2, P3, P4 are the Process names here 
	// Allocation Matrix 
	alloc = new int[][] { { 0, 1, 0 }, //P0 
				{ 2, 0, 0 }, //P1 
				{ 3, 0, 2 }, //P2 
				{ 2, 1, 1 }, //P3 
				{ 0, 0, 2 } }; //P4 
			
	// MAX Matrix 
	max = new int[][] { { 7, 5, 3 }, //P0 
			{ 3, 2, 2 }, //P1 
			{ 9, 0, 2 }, //P2 
			{ 2, 2, 2 }, //P3 
			{ 4, 3, 3 } }; //P4 
	
	// Available Resources 
	avail = new int[] { 3, 3, 2 }; 
	} 

	void isSafe() 
	{ 
	int count=0; 
	
	//visited array to find the already allocated process 
	boolean visited[] = new boolean[n]; 
	for (int i = 0;i < n; i++) 
	{ 
		visited[i] = false; 
	} 
		
	//work array to store the copy of available resources 
	int work[] = new int[m];	 
	for (int i = 0;i < m; i++) 
	{ 
		work[i] = avail[i]; 
	} 

	while (count<n) 
	{ 
		boolean flag = false; 
		for (int i = 0;i < n; i++) 
		{ 
			if (visited[i] == false) 
			{ 
			int j; 
			for (j = 0;j < m; j++) 
			{		 
				if (need[i][j] > work[j]) 
				break; 
			} 
			if (j == m) 
			{ 
			safeSequence[count++]=i; 
			visited[i]=true; 
			flag=true; 
						
				for (j = 0;j < m; j++) 
				{ 
				work[j] = work[j]+alloc[i][j]; 
				} 
			} 
			} 
		} 
		if (flag == false) 
		{ 
			break; 
		} 
	} 
	if (count < n) 
	{ 
		System.out.println("The System is UnSafe!"); 
	} 
	else
	{ 
		//System.out.println("The given System is Safe"); 
		System.out.println("Following is the SAFE Sequence"); 
				for (int i = 0;i < n; i++) 
		{ 
			System.out.print("P" + safeSequence[i]); 
			if (i != n-1) 
			System.out.print(" -> "); 
		} 
	} 
	} 
	
	void calculateNeed() 
	{ 
	for (int i = 0;i < n; i++) 
	{ 
		for (int j = 0;j < m; j++) 
		{ 
		need[i][j] = max[i][j]-alloc[i][j]; 
		} 
	}		 
	} 

	public static void main(String[] args) 
	{ 
	int i, j, k; 
	Bankers gfg = new Bankers(); 
		
	gfg.initializeValues();	 
	//Calculate the Need Matrix	 
	gfg.calculateNeed();			 
			
	// Check whether system is in safe state or not 
	gfg.isSafe();		 
	} 
} 

File: algorithms/Bankers_Algorithm_Operating_System/Bankers.c

# Number of processes 
P = 5

# Number of resources 
R = 3

# Function to find the need of each process 
def calculateNeed(need, maxm, allot): 

	# Calculating Need of each P 
	for i in range(P): 
		for j in range(R): 
			
			# Need of instance = maxm instance - 
			# allocated instance 
			need[i][j] = maxm[i][j] - allot[i][j] 

# Function to find the system is in 
# safe state or not 
def isSafe(processes, avail, maxm, allot): 
	need = [] 
	for i in range(P): 
		l = [] 
		for j in range(R): 
			l.append(0) 
		need.append(l) 
		
	# Function to calculate need matrix 
	calculateNeed(need, maxm, allot) 

	# Mark all processes as infinish 
	finish = [0] * P 
	
	# To store safe sequence 
	safeSeq = [0] * P 

	# Make a copy of available resources 
	work = [0] * R 
	for i in range(R): 
		work[i] = avail[i] 

	# While all processes are not finished 
	# or system is not in safe state. 
	count = 0
	while (count < P): 
		
		# Find a process which is not finish 
		# and whose needs can be satisfied 
		# with current work[] resources. 
		found = False
		for p in range(P): 
		
			# First check if a process is finished, 
			# if no, go for next condition 
			if (finish[p] == 0): 
			
				# Check if for all resources 
				# of current P need is less 
				# than work 
				for j in range(R): 
					if (need[p][j] > work[j]): 
						break
					
				# If all needs of p were satisfied. 
				if (j == R - 1): 
				
					# Add the allocated resources of 
					# current P to the available/work 
					# resources i.e.free the resources 
					for k in range(R): 
						work[k] += allot[p][k] 

					# Add this process to safe sequence. 
					safeSeq[count] = p 
					count += 1

					# Mark this p as finished 
					finish[p] = 1

					found = True
				
		# If we could not find a next process 
		# in safe sequence. 
		if (found == False): 
			print("System is not in safe state") 
			return False
		
	# If system is in safe state then 
	# safe sequence will be as below 
	print("System is in safe state.", 
			"\nSafe sequence is: ", end = " ") 
	print(*safeSeq) 

	return True

# Driver code 
if __name__ =="__main__": 
	
	processes = [0, 1, 2, 3, 4] 

	# Available instances of resources 
	avail = [3, 3, 2] 

	# Maximum R that can be allocated 
	# to processes 
	maxm = [[7, 5, 3], [3, 2, 2], 
			[9, 0, 2], [2, 2, 2], 
			[4, 3, 3]] 

	# Resources allocated to processes 
	allot = [[0, 1, 0], [2, 0, 0], 
			[3, 0, 2], [2, 1, 1], 
			[0, 0, 2]] 

	# Check system is in safe state or not 
	isSafe(processes, avail, maxm, allot) 
